### PR TITLE
RFC: pex-based in-place packages.

### PR DESCRIPTION
**Pull request by mikekap**
This adds support for another python package style - pex_inplace. It uses the regular pex code to make the chroot, but then just doesn't zip it. This is a use case supported by pex.

Also included are some optimizations to pex generation, but the speedup is really only seen with the in-place version. Generating buck.pex via regular pex takes ~ 4-5s on my mac; with pex_inplace it's down to ~300ms. This is still a little slower than the regular inplace mode, but supports more features (e.g. prebuilt packages, proper isolation, etc).

The optimizations to pexing are:
 - Don't produce a code hash in PEX-INFO in inplace mode. This is only used to check if we need to extract the contents of a pex from a zip; if it's already a directory, it's useless. (3-4s)
 - Pass the python version information from buck to pex.py instead of querying it every time make_pex is run (400ms).

### CODE REVIEW

**Comment by mikekap** [comment]
ping :)

**Comment by sdwilsh** [comment]
I think @andrewjcg is on a plane today, but he might be jet lagged enough to look at this when he lands :)

**Comment by mikekap** [comment]
Sorry I didn't mean to rush, just wanted to make sure this didn't get dropped given the pex commit/revert stuff.

**Comment by mikekap** [comment]
Ping :)

**Comment by mikekap** [comment]
If it'll help, I can split this up into two changes: one to add pex_inplace and the other to optimize the ~400ms interpreter lookup.

**Comment by sdwilsh** [comment]
@facebook-github-bot import

**Comment by andrewjcg** [comment]
So sorry for taking this long to get to this, but it's definitely on my radar now.  Generally it looks good, I just have a few concerns and suggestions about separating out the PEX-specific stuff from the more general python stuff which I'll comment on inline.

File: src/com/facebook/buck/python/AbstractPythonVersion.java [hunk 1]
```diff
@@ -25,11 +25,22 @@
...
+  @Value.Parameter
+  public abstract String getVersionString();  // X.Y.Z
+
+  public String getPexCompatibilityVersion() {
+    StringBuilder sb = new StringBuilder(20);
+    sb.append(getInterpreterName());
+    sb.append(" ");
+    sb.append(getVersionString().replace('.', ' '));
+    return sb.toString();
+  }
```
Comment (andrewjcg) [inline src/com/facebook/buck/python/AbstractPythonVersion.java]: It'd be nice if we didn't have more PEX-specific stuff in the core python support.  Can we just have `PythonVersion` support the interpreter-name and version (as you've added above) and have the make_pex.py script do the necessary PEX-specific formatting?

File: src/com/facebook/buck/python/PythonBuckConfig.java [hunk 2]
```diff
@@ -286,6 +301,7 @@ public PackageStyle getPackageStyle() {
   public enum PackageStyle {
     STANDALONE,
     INPLACE,
+    PEX_INPLACE,
```
Comment (andrewjcg) [inline src/com/facebook/buck/python/PythonBuckConfig.java]: Long-term we'd like to move all the PEX-specific naming/code out of Buck, just supporting the general way to use arbitrary custom Python packagers via the `python:path_to_pex` config setting.

In this vein, it'd be nice to at least keep the zip-vs-directory choice separate from the inplace-vs-standalone choice. What do you think about having a separate PexStyle enum with `ZIP` and `DIRECTORY` choices (which would apply when `STANDALONE` is used with the default PEX builder)?

Another option might be to make the current in-place package style more general and provide a config option to choose between "simple" and "PEX" sub-styles, the latter of which implements the in-place PEX. The annoying thing about this is that the current code path for in-place assumes that the generated symlink trees with the modules will be used and that the in-place starter script won't do any of it's own packaging (even if it is just copying to a directory).

Comment (mikekap) [inline src/com/facebook/buck/python/PythonBuckConfig.java]: Hmm I think if the goal is to remove most of the pex-specific parts of the code, I think just adding something like `python:pex_args` and using the existing python platform flavor machinery might be the right approach. At that point I think it would make sense to convert the `INPLACE` style to this scheme, by e.g. having a magic `path_to_pex` value that uses the existing inplace code. Alternatively, an option might be to extract the existing inplace code into a java binary, although that might be a bit more of a rewrite.

**Comment by sdwilsh** [comment]
@facebook-github-bot import

File: test/com/facebook/buck/python/PythonBinaryIntegrationTest.java [hunk 3]
```diff
@@ -126,6 +141,10 @@ public void commandLineArgs() throws IOException {
     assertThat(
         result.getStdout(),
         containsString("HELLO WORLD"));
+
+    if (pexDirectory) {
```
Comment (andrewjcg) [inline test/com/facebook/buck/python/PythonBinaryIntegrationTest.java]: It seems a bit weird to test for this in the `commandLineArgs` test.  Maybe move this to a separate `testOutput` test where we check for a directory for the `--directory` case and files for the others?

File: test/com/facebook/buck/python/PythonBinaryIntegrationTest.java [hunk 4]
```diff
@@ -287,10 +306,10 @@ private PythonBuckConfig getPythonBuckConfig() throws IOException {
             new ProjectFilesystem(tmp.getRootPath()),
             Architecture.detect(),
             Platform.detect(),
-            ImmutableMap.<String, String>of());
+            ImmutableMap.copyOf(System.getenv()));
```
Comment (andrewjcg) [inline test/com/facebook/buck/python/PythonBinaryIntegrationTest.java]: Why was this necessary?

Comment (andrewjcg) [inline test/com/facebook/buck/python/PythonBinaryIntegrationTest.java]: Was this to find the `python` binary in the path?

**Comment by andrewjcg** [comment]
Looks good thanks!  Just a few minor questions/comments inline.

File: test/com/facebook/buck/python/PythonBinaryIntegrationTest.java (hunk 4 above)
Comment (mikekap) [inline test/com/facebook/buck/python/PythonBinaryIntegrationTest.java]: Yep. I think this might also help windows tests since you can't use shebangs.

**Comment by mikekap** [comment]
Ping

**Comment by Coneko** [comment]
I'm going to import since @andrewjcg said the remaining comments were minor and they seem to have been addressed.

**Comment by Coneko** [comment]
@facebook-github-bot shipit

**Comment by sdwilsh** [comment]
It looks like @andrewjcg reverted this in https://github.com/facebook/buck/commit/a03737b3e2e31e833e30e9013052dad7a936e5cb, but I don't see a comment here about it.
